lib/jxl/enc_frame.cc

```diff
@@ -1572,39 +1629,307 @@ bool CanDoStreamingEncoding(const CompressParams& cparams,
+  size_t bucket = 0;
+  while (bucket < 3 && group_size >= kGroupSizeOffset[bucket + 1]) ++bucket;
+  return bucket;
+}
+
+size_t TOCSize(const std::vector<size_t>& group_sizes) {
+  size_t toc_bits = 0;
+  for (size_t i = 0; i < group_sizes.size(); i++) {
+    toc_bits += kTOCBits[TOCBucket(group_sizes[i])];
+  }
+  return (toc_bits + 7) / 8;
+}
+
+PaddedBytes EncodeTOC(const std::vector<size_t>& group_sizes, AuxOut* aux_out) {
+  BitWriter writer;
+  BitWriter::Allotment allotment(&writer, 32 * group_sizes.size());
+  for (size_t i = 0; i < group_sizes.size(); i++) {
+    JXL_CHECK(U32Coder::Write(kTocDist, group_sizes[i], &writer));
+  }
+  writer.ZeroPadToByte();  // before first group
+  allotment.ReclaimAndCharge(&writer, kLayerTOC, aux_out);
+  return std::move(writer).TakeBytes();
+}
+
+void ComputeGroupDataOffset(size_t frame_header_size, size_t dc_global_size,
+                            size_t num_sections, size_t& min_dc_global_size,
+                            size_t& group_offset) {
+  size_t max_toc_bits = (num_sections - 1) * 32;
+  size_t min_toc_bits = (num_sections - 1) * 12;
+  size_t max_padding = (max_toc_bits - min_toc_bits + 7) / 8;
+  min_dc_global_size = dc_global_size;
+  size_t dc_global_bucket = TOCBucket(min_dc_global_size);
+  while (TOCBucket(min_dc_global_size + max_padding) > dc_global_bucket) {
+    dc_global_bucket = TOCBucket(min_dc_global_size + max_padding);
+    min_dc_global_size = kGroupSizeOffset[dc_global_bucket];
+  }
+  JXL_ASSERT(TOCBucket(min_dc_global_size) == dc_global_bucket);
+  JXL_ASSERT(TOCBucket(min_dc_global_size + max_padding) == dc_global_bucket);
+  max_toc_bits += kTOCBits[dc_global_bucket];
+  size_t max_toc_size = (max_toc_bits + 7) / 8;
+  group_offset = frame_header_size + max_toc_size + min_dc_global_size;
+}
+
+size_t ComputeDcGlobalPadding(const std::vector<size_t>& group_sizes,
+                              size_t frame_header_size,
+                              size_t group_data_offset,
+                              size_t min_dc_global_size) {
+  std::vector<size_t> new_group_sizes = group_sizes;
+  new_group_sizes[0] = min_dc_global_size;
+  size_t toc_size = TOCSize(new_group_sizes);
+  size_t actual_offset = frame_header_size + toc_size + group_sizes[0];
+  return group_data_offset - actual_offset;
+}
+
+Status OutputGroups(std::vector<BitWriter>&& group_codes,
+                    std::vector<size_t>* group_sizes,
+                    JxlEncoderOutputProcessorWrapper* output_processor) {
+  JXL_ASSERT(group_codes.size() >= 4);
+  {
+    PaddedBytes dc_group = std::move(group_codes[1]).TakeBytes();
+    group_sizes->push_back(dc_group.size());
+    JXL_RETURN_IF_ERROR(AppendData(*output_processor, dc_group));
+  }
+  for (size_t i = 3; i < group_codes.size(); ++i) {
+    PaddedBytes ac_group = std::move(group_codes[i]).TakeBytes();
+    group_sizes->push_back(ac_group.size());
+    JXL_RETURN_IF_ERROR(AppendData(*output_processor, ac_group));
+  }
+  return true;
+}
```

Comment:
I'll assume that pretty much all the code from `kGroupSizeOffset` to here was taken from the fast-lossless code - can we move that to some shared file? (that code would not run in standalone mode anyway)

Reply:
The table lookups are used in the standalone mode to write the actual TOC, and the other functions are all slightly different. E.g. here we know the final frame header because we call this whole EncodeFrame() in the output part, while in fast lossless we don't because we call that in the input part to avoid an extra memcpy for the non-streaming input case. Also the fast lossless code has no DC groups so the padding logic is also a bit different.

Reply:
Probably still possible to make it a shared file, but we can do that as a follow-up